rohmu/object_storage/base.py

```diff
@@ -175,29 +206,71 @@ def sanitize_metadata(self, metadata, replace_hyphen_with="-"):
         """Convert non-string metadata values to strings and drop null values"""
         return {str(k).replace("-", replace_hyphen_with): str(v) for k, v in (metadata or {}).items() if v is not None}
 
-    def store_file_from_memory(self, key, memstring, metadata=None, cache_control=None, mimetype=None):
-        raise NotImplementedError
+    def store_file_from_memory(
+        self,
+        key,
+        memstring,
+        metadata=None,
+        *,
+        cache_control=None,
+        mimetype=None,
+        multipart: Union[bool, None] = None,
+        progress_fn: ProgressProportionCallbackType = None,
+    ):
+        with BytesIO(memstring) as buf:
+            size = len(memstring)
+            if metadata is None:
+                metadata = {"Content-Length": size}
+            elif metadata.get("Content-Length") is None:
+                metadata = metadata.copy()
+                metadata["Content-Length"] = size
+            self.store_file_object(
+                key,
+                buf,
+                cache_control=cache_control,
+                metadata=metadata,
+                mimetype=mimetype,
+                multipart=multipart,
+                upload_progress_fn=self._incremental_to_proportional_progress(cb=progress_fn, size=size),
+            )
 
     def store_file_from_disk(
         self,
         key,
         filepath,
         metadata=None,
-        multipart=None,
+        *,
         cache_control=None,
         mimetype=None,
+        multipart: Union[bool, None] = None,
         progress_fn: ProgressProportionCallbackType = None,
     ):
-        raise NotImplementedError
+        size = os.path.getsize(filepath)
+        with open(filepath, "rb") as fd:
+            if metadata is None:
+                metadata = {"Content-Length": size}
+            elif metadata.get("Content-Length") is None:
+                metadata = metadata.copy()
+                metadata["Content-Length"] = size
+            self.store_file_object(
+                key,
+                fd,
+                cache_control=cache_control,
+                metadata=metadata,
+                mimetype=mimetype,
+                multipart=multipart,
+                upload_progress_fn=self._incremental_to_proportional_progress(cb=progress_fn, size=size),
```

Comment:
This looks to be incorrect behavior for `S3Transfer` as that reports the total number of bytes sent every time instead of delta.